### ISSUE TITLE
Capture exceptions and expose them to the REPL

### ISSUE THREAD

**Issue by lihaoyi** [lihaoyi/Ammonite#289]
EDIT: Original issue was for `ThreadDeath`, but this will be useful for any exception

This can be exposed to the REPL as a value, and would be useful for a variety of reasons:

- You may want to know "how far" execution went before you interrupted it. Inspecting the stack trace would give you some indication of "where" in the executed code it had reached.

- A prime case of interrupting code is when it infinite-loops; in such a case you'd like to know where the infinite loop is! Since that's where the program is spending "all" its time, the stack trace (perhaps repeated once or twice) should give a very good indication of where the program is getting stuck

- If the program is blocked/deadlocked on something indefinitely when you interrupt it (e.g. out of frustration) the stack trace will point directly to the offending line of code.

Currently it just prints 
```scala
haoyi-static-analysis@ while(true)()
^C
Interrupted!
```
When you interrupt it. We could instead assign it to some value
```scala
haoyi-static-analysis@ while(true)()
^C
val res12: InterruptedException = ...
```

Which would both show the user that it was interrupted, and let them dig deeper if they wished to

**Comment by lihaoyi** [comment]
Actually this might not be unique to ThreadDeath; any exception that gets thrown should be made available as a `val`, as sometimes they contain useful metadata e.g. FastParse's `ParseError`s or Ammonite-Ops's `ShelloutException`s. 

By making these uncaught exceptions available to the shell session by default, we could save a lot of annoying re-running-code-with-try-catch during interactive development and exploration

**Comment by lihaoyi** [comment]
Turn out the normal REPL has a mutable `lastException: Throwable` definition. This is halfway there, but isn't ideal: you still need to cast the `Throwable` into the correct type before being able to manipulate it usefully. The REPL should already know what the type is, and by exposing it as a `val` of the correct type, you won't need to do the casting yourself.

**Comment by lihaoyi** [comment]
This has been fixed in 0.5.5; thanks @coderabhishek !
